armsr: fix serial console regression on device tree systems

mcbridematt opened the pull request:
~This reverts a portion of #16213 . The addition to inittab for tty1 and change of cmdline on x86 remains.~
**New (January 2025) proposed solution: https://github.com/openwrt/openwrt/pull/17012#issuecomment-2591751115**

This solution makes the kernel always setup the framebuffer (tty0) console, preventing the need to specify any console= on the command line in both serial and screen cases.
The kernel was not setting up tty0 when an ACPI-specified serial port (SPCR) is present, as some system firmwares do not remove the SPCR when a screen is present.

----  


The addition of console=tty1 to the kernel command line will break the default console setup on boards using device tree and without a graphical/framebuffer console, as that will replace any setting supplied by the system firmware.

Instead of forcing the console=tty1 as default on all armsr images, I suggest we do:

1. Add a separate image profile intended for graphical setups (the generic image will work on all machines, but with the kernel boot output not visible on framebuffers)

and/or

**2.** Patch the kernel so the "stdout" console specified in device tree is activated alongside any console= specified on the command line, mirroring existing behavior on ACPI systems (see below) **(current proposed solution)**

and/or

3. Set up some method of persisting custom command lines across sysupgrade (e.g something in /etc/config gets applied by sysupgrade, like `/etc/default/grub` on mainstream distros), so if the user wants to customize the setting for their specific system, it won't be overriden on upgrade.


This is what happens on my LS1088A board:
```
[    0.000000] Linux version 6.6.60 (buildbot@bf03311b041f) (aarch64-openwrt-linux-musl-gcc (OpenWrt GCC 13.3.0 r27587+508-38bb47c36c) 13.3.0, GNU ld (GNU Binutils) 2.42) #0 SMP Fri Nov 15 03:31:56 2024
[    0.000000] KASLR enabled
[    0.000000] Machine model: Traverse Ten64
...
[    0.008413] Console: colour dummy device 80x25
[    0.012882] printk: console [tty1] enabled
[    0.017000] printk: bootconsole [ns16550a0] disabled
< no further output >
```

The tty on ttyS0 is spawned, but at 9600 instead of the 115200 the board started with.
I have reproduced the issue on another SoC (the Coral 'phanbell' dev board / NXP i.MX8M+). I believe most other boards using device tree and no framebuffer's will also be affected.

When the QEMU virtual machine is used with u-boot and no graphics (`-nographic`) it is also slightly affected, with no kernel output between the serial console being disabled and askfirst:
```
[    0.004485] printk: console [tty1] enabled
[    0.005919] printk: bootconsole [pl11] disabled
Please press Enter to activate this console.
```
The QEMU console will adopt the baud rate that the virtualized OS sets up, meaning the lack of baud rate retention is not a visible issue.

Boards with ACPI firmware do not appear to be affected as the kernel will activate consoles specified by ACPI (in the SPCR table) regardless of the console= supplied.
On DT we have the `stdout-path` attribute under the `chosen` node, but any `console=` settings will override it.

ynezz commented:
> This reverts a portion of https://github.com/openwrt/openwrt/pull/16213 . The addition to inittab for tty1 and change of cmdline on x86 remains.

@markmentovai FYI

markmentovai commented:
This patch predictably breaks the framebuffer console, which is what #16213 was intended to fix.

I didn’t expect that #16213 would break the serial console. In fact, I run with both a framebuffer and serial console, and found both to function properly after #16213 (without this patch):

```
# cat /proc/cmdline 
BOOT_IMAGE=/efi/openwrt/vmlinux root=PARTUUID=… rootwait earlycon console=tty1 noinitrd
# cat /proc/consoles
tty1                 -WU (EC    )    4:1
ttyAMA0              -W- (E  p a)  204:64
# dmesg | grep -i console
[    0.000000] ACPI: SPCR: console: pl011,mmio32,0x9000000,9600
[    0.000000] printk: bootconsole [pl11] enabled
[    0.000000] Kernel command line: BOOT_IMAGE=/efi/openwrt/vmlinux root=… rootwait earlycon console=tty1 noinitrd
[    0.001708] Console: colour dummy device 80x25
[    0.002676] printk: console [tty1] enabled
[    0.003849] printk: bootconsole [pl11] disabled
[    0.017682] printk: console [ttyAMA0] enabled
[    0.423332] Console: switching to colour frame buffer device 160x50
[    0.561832] init: Console is alive
```

With this patch, only the serial console works:

```
# cat /proc/cmdline
BOOT_IMAGE=/efi/openwrt/vmlinux root=PARTUUID=… rootwait earlycon noinitrd
# cat /proc/consoles
ttyAMA0               -W- (EC   a)  204:64
# dmesg | grep -i console
[    0.000000] ACPI: SPCR: console: pl011,mmio32,0x9000000,9600
[    0.000000] printk: bootconsole [pl11] enabled
[    0.002702] Console: colour dummy device 80x25
[    0.110466] printk: console [ttyAMA0] enabled
[    0.112270] printk: bootconsole [pl11] disabled
[    0.314638] Console: switching to colour frame buffer device 160x50
[    0.428738] init: Console is alive
```

So, sorry for breaking your use. I did test and didn’t consider that implicit consoles coming in from DT would behave differently from those coming in from ACPI.

For what it’s worth, this portion of #16213 was attempting to make armsr (armv8) work more like x86 (x86_64), where I’m able deploy the same image to physical machines with only serial consoles, as well as to virtual machines with (usually) only framebuffer consoles, and both work correctly. It’s incredibly valuable to be able to use the same image both virtualized and not virtualized. I would prefer not to abandon that property.

For reference, x86_64 (whether virtualized or not):

```
# cat /proc/cmdline
BOOT_IMAGE=/boot/vmlinuz root=PARTUUID=… rootwait console=tty1 console=ttyS0,115200n8 noinitrd
# cat /proc/consoles
ttyS0                -W- (EC p a)    4:64
tty1                 -WU (E  p  )    4:1
# dmesg|grep -i console
[    0.000000] Command line: BOOT_IMAGE=/boot/vmlinuz root=… rootwait console=tty1 console=ttyS0,115200n8 noinitrd
[    0.000000] Kernel command line: BOOT_IMAGE=/boot/vmlinuz root=PARTUUID=… rootwait console=tty1 console=ttyS0,115200n8 noinitrd
[    0.000000] Console: colour dummy device 80x25
[    0.000000] printk: console [tty1] enabled
[    0.000000] printk: console [ttyS0] enabled
[    0.870136] Console: switching to colour frame buffer device 160x50
[    2.263414] init: Console is alive
```

So I hear what you’re saying, and think that something like option 2 may be most palatable. It would be best if the console appeared on both serial port and framebuffer, as they exist, in as many situations as possible.

I’d prefer to avoid a situation where my patch breaks your use case and your patch makes my use case (which is what this does as it stands now). Can we find something that works more universally?

x86’s command line has `console=` naming both the framebuffer and serial port. Would it maybe make sense to allow for an explicit `console=keep` option that would prevent subsequent `console=` arguments from overriding what was specified by firmware? Alternatively, making the DT case behave more like the ACPI case, where `console=` doesn’t override the implicit incoming console, could make sense (your option 2).

mcbridematt commented:
> I’d prefer to avoid a situation where my patch breaks your use case and your patch makes my use case (which is what this does as it stands now). Can we find something that works more universally?

I agree! I'd like it to "just work" everywhere, preferably without different command lines or image profiles.

So I worked out that the reason why the kernel activates the provided serial on ACPI but not DT, is that the DT/OF init sequence checks for `console=` on the command line and _only_ then calls `add_preferred_console` - see [drivers/of/base.c#L1919](https://github.com/torvalds/linux/blob/9f16d5e6f220661f73b36a4be1b21575651d8833/drivers/of/base.c#L1919).
For reference, the ACPI equivalent is in [drivers/acpi/spcr.c#L223](https://github.com/torvalds/linux/blob/9f16d5e6f220661f73b36a4be1b21575651d8833/drivers/acpi/spcr.c#L223)

Remove that check and the serial console is always setup.
(I was worried there would be a lot more work involved to implement that!)

I think that will fix the problem at hand and allow serial ttys to continue working in the presence of any console= argument present.
Does anyone have any concerns with this approach?

ynezz commented:
> So I worked out that the reason why the kernel activates the provided serial on ACPI but not DT,

Nice, thanks!

>  Does anyone have any concerns with this approach?

What approach exactly? :) Are you going to suggest removing this check upstream or keep this as a downstream patch? If the former, I like this approach.

ynezz commented:
> What approach exactly? :) Are you going to suggest removing this check upstream or keep this as a downstream patch? If the former, I like this approach.

Ok, just found `(This patch is not intended for upstream submission)`, so downstream workaround, can't we really do better?

> Kernel patches for armsr/armv7 require refresh. (run 'make target/linux/refresh' and force push this pr)

https://github.com/openwrt/openwrt/actions/runs/12003987975/artifacts/2231433838

mcbridematt commented:
> Kernel patches for armsr/armv7 require refresh. (run 'make target/linux/refresh' and force push this pr)

Oops, sorry. Refreshed the patches and pushed again.

> can't we really do better?

I think a KConfig gate/setting (at a minimum) would be needed to send that upstream, and it will likely need a way to disable that feature on the command line. I've discovered there was recently a [acpi=nospcr](https://patchwork.kernel.org/project/linux-arm-kernel/patch/20240625030504.58025-1-liuwei09@cestc.cn/) command line added to negate the ACPI behavior.

It would be nicer if we didn't need `console=ttyX` to force printk to appear on the framebuffer as well. I think I need to keep searching, there must be a KConfig somewhere to make that happen..

mcbridematt commented:
Apologies for the long delay, I took a break from coding over the (Australian) summer holidays.

I have taken a fresh look at this and I think I have a better solution. I would like to get this solved before the OpenWrt 24.10 release.

> It would be nicer if we didn't need console=ttyX to force printk to appear on the framebuffer as well. I think I need to keep searching, there must be a KConfig somewhere to make that happen..

The answer is that Linux will setup a console on `tty0` by default*, but only if the kernel has not found a "preferred" console by the time kernel sets up a dummy console device:
```
Console: colour dummy device 80x25
```
This dummy tty0 device is handed over to the "real" framebuffer, when it is ready.

(* assuming CONFIG_VT_CONSOLE and CONFIG_DUMMY_CONSOLE are enabled)

As mentioned before, on ACPI systems with an SPCR table, the parsing of the SPCR happens very early, so the preferred console list already has an entry:
```
root@OpenWrt:~# dmesg | grep -E '(SPCR|dummy device|__add_preferred_console|tty)'
[    0.000000] ACPI: SPCR 0x00000000784B0000 000050 (v02 BOCHS  BXPC     00000001 BXPC 00000001)
[    0.000000] ACPI: SPCR: console: pl011,mmio,0x9000000,9600
[    0.000000] printk: __add_preferred_console: name=pl011 idx=0 user_specified=0
[    0.002855] Console: colour dummy device 80x25
[    0.112606] ARMH0011:00: ttyAMA0 at MMIO 0x9000000 (irq = 12, base_baud = 0) is a SBSA
[    0.115392] printk: console [ttyAMA0] enabled
[    0.468220] 0000:00:06.0: ttyS0 at I/O 0x1160 (irq = 47, base_baud = 115200) is a 16550A
root@OpenWrt:~# cat /proc/consoles
ttyAMA0              -W- (EC   a)  204:64
```

On DT, the equivalent (stdout-path) isn't parsed until later, so tty0 is setup as console:

```
[    0.003018] Console: colour dummy device 80x25
[    0.004770] printk: console [tty0] enabled
[    0.021351] printk: __add_preferred_console: name=ttyAMA idx=0 user_specified=0
[    0.021365] 9000000.pl011: ttyAMA0 at MMIO 0x9000000 (irq = 13, base_baud = 0) is a PL011 rev1
[    0.021472] printk: console [ttyAMA0] enabled
[    0.712773] 0000:00:06.0: ttyS0 at I/O 0x1160 (irq = 15, base_baud = 115200) is a 16550A
# cat /proc/consoles
ttyAMA0              -W- (EC p a)  204:64
tty0                 -WU (E     )    4:1

```

So when running QEMU with virtio-gpu and DT bootloader, we already get consoles on the screen and serial port without having to add any console= options.

We can make this happen by circumventing the `preferred_console < 0` check in printk.c, forcing a call to `try_enable_default_console` which sets tty0 as a console device:
https://github.com/torvalds/linux/blob/619f0b6fad524f08d493a98d55bac9ab8895e3a6/kernel/printk/printk.c#L4003-L4008

For our purposes, I propose changing this so the tty0 activation occurs whenever there is no `console=` on the kernel cmdline:
```
	if (!console_set_on_cmdline) {
		if (hlist_empty(&console_list) || !console_first()->device ||
		    console_first()->flags & CON_BOOT) {
			try_enable_default_console(newcon);
		}
	} 
```

Which will give us our desired result (console on screen + serial) without any console= arguments on both ACPI and DT:
```
# ACPI
root@OpenWrt:~# dmesg | grep -E '(SPCR|dummy device|__add_preferred_console|tty)'
[    0.000000] ACPI: SPCR 0x00000000784B0000 000050 (v02 BOCHS  BXPC     00000001 BXPC 00000001)
[    0.000000] ACPI: SPCR: console: pl011,mmio,0x9000000,9600
[    0.000000] printk: __add_preferred_console: name=pl011 idx=0 user_specified=0
[    0.002927] Console: colour dummy device 80x25
[    0.004622] printk: console [tty0] enabled
[    0.025189] ARMH0011:00: ttyAMA0 at MMIO 0x9000000 (irq = 12, base_baud = 0) is a SBSA
[    0.025210] printk: console [ttyAMA0] enabled
[    0.596424] 0000:00:06.0: ttyS0 at I/O 0x1160 (irq = 47, base_baud = 115200) is a 16550A
root@OpenWrt:~# cat /proc/consoles
ttyAMA0              -W- (EC p a)  204:64
tty0                 -WU (E     )    4:1
root@OpenWrt:~# cat /proc/cmdline
BOOT_IMAGE=/efi/openwrt/vmlinux root=PARTUUID=d1f37050-1e52-ffc8-494f-54c543df6202 rootwait earlycon noinitrd

# DT
[    0.002939] Console: colour dummy device 80x25
[    0.004628] printk: console [tty0] enabled
[    0.021253] printk: __add_preferred_console: name=ttyAMA idx=0 user_specified=0
[    0.021266] 9000000.pl011: ttyAMA0 at MMIO 0x9000000 (irq = 13, base_baud = 0) is a PL011 rev1
[    0.021375] printk: console [ttyAMA0] enabled
[    0.696120] 0000:00:06.0: ttyS0 at I/O 0x1160 (irq = 15, base_baud = 115200) is a 16550A
root@OpenWrt:~# cat /proc/consoles
ttyAMA0              -W- (EC p a)  204:64
tty0                 -WU (E     )    4:1
```

This is still a hack, but (IMO) a much cleaner one, and if you want a console only on a specific device, you can specify one through the kernel command line.
If there are no objections, I'll prepare a new patch set in the next couple of days.

The alternative would be to add support for something like `console=dt-stdout`, but I think that would involve a lot more changes to printk.c, and unlikely to lead to an upstream solution.

mcbridematt committed:
armsr: add patch (hack) to enable all "default" consoles

A previous change added 'console=tty1' to the default kernel command
line on armsr, in order to ensure the framebuffer console is enabled
on systems capable of graphics output.

Unfortunately, this change broke boards that used device tree
(DT) firmware with serial consoles, as the serial console
specified by the system firmware (stdout-path) was no longer
setup by the kernel.

A bit of probing determined that the SPCR (serial port console
direction table) on ACPI systems was preventing Linux from setting
up a default framebuffer console on these systems (which is why
console=tty1 was added).

(The affected ACPI systems are usually VMs using QEMU's
'virt' machine and EDK2 firmware. The firmware on these systems
does not remove the SPCR when a screen is present)

So to ensure all possible systems are setup correctly, we modify
the kernel so all "default" console types (serial and screen)
are setup when no console= arguments are specified on the kernel
command line.

Signed-off-by: Mathew McBride <matt@traverse.com.au>
Fixes: c099523d66a3 ("use console=tty1 to make console more
readily available")
Link: https://github.com/openwrt/openwrt/pull/17012
Signed-off-by: Robert Marko <robimarko@gmail.com>

mcbridematt committed:
armsr: remove 'console=tty1' from kernel command line

We have modified the kernel to setup all "default" consoles,
including serial ports and framebuffers/screens, providing
no console= argument is supplied on the kernel command line.

Adding 'console=tty1' caused the 'default' serial port on
device tree systems to break, as the kernel would not carry
over the settings (like baud rate) from the bootloader.

The system administrator can still force the use of a
specific console by adding their own console= arguments.

Signed-off-by: Mathew McBride <matt@traverse.com.au>
Fixes: c099523d66a3 ("armsr: use console=tty1 to make
console more readily available")
Link: https://github.com/openwrt/openwrt/pull/17012
Signed-off-by: Robert Marko <robimarko@gmail.com>

mcbridematt commented:
@markmentovai Can you try the changes above and see if the consoles work correctly under your setup?
It is slightly different in that tty0 will be the main "screen" console instead of tty1, but it should function the same (unless I have missed something?)
Pre-built images here: https://archive.traverse.com.au/pub/traverse/ls1088firmware/openwrt/branches/openwrt-master-sr-6-6/1627383878/image/

I have tested these configurations so far and each works correctly
QEMU + UEFI/EDK2 (ACPI) with virtio-gpu
QEMU + u-boot (DT) with virtio-gpu
Ten64 (u-boot+DT), serial console only
Windows HyperV on Arm64/SnapdragonX (ACPI) (doesn't fully work yet, but the kernel messages display correctly)

markmentovai commented:
Thanks for looking into this in more detail, @mcbridematt! I tested the combination of https://github.com/openwrt/openwrt/pull/17012/commits/2e7cfeb70c8eb586623f5e09b572a437d474f02a and https://github.com/openwrt/openwrt/pull/17012/commits/66727948aff01ff81cd65b13509716594d9ab232, and it works for me.

robimarko commented:
Any chance you can rebase this to also add 6.12 patch?

mcbridematt commented:
> Any chance you can rebase this to also add 6.12 patch?

Would it be better to get this pull in first, so kernel_bump can carry the patch across to 6.12?

The armsr/6.12 pull request now contains a refresh of the patch: https://github.com/openwrt/openwrt/pull/18849/commits/c0f808b21b170dce51d000a785097807445d6f51
